agency.rb

```diff
@@ -88,13 +88,14 @@ def create_managers
     # This method will block until all middleware instances are fully
     # initialized (their `ready?` method returns true).
     def create_middlewares
-      @middlewares = self.class.middlewares.each.with_object(nil).map do |(klass, args, kwargs), app|
-        klass.new(app, *args, **kwargs)
+      instance_list = []
+      @middleware = self.class.middlewares.inject(nil) do |app, (klass, args, kwargs)|
+        klass.new(app, *args, **kwargs).tap{ |inst| instance_list << inst }
       end
       # Some Middlewares will use background threads to process work. By
       # sleeping for a short time between checks, those threads can work
       # concurrently instead of each one blocking serially.
-      sleep(0.01) until @middlewares.drop_while(&:ready?).empty?
+      sleep(0.01) until instance_list.drop_while(&:ready?).empty?
     end
```

Comment:
This commit also fixes the way Middlewares are initialized (since there is now a stack of middleware rather than just one).

Before, `app` was always set to `nil` because `each.with_object` does not take the return value of the previous iteration, but rather keeps passing the same object (`nil` here).